program/rust/src/tests/pyth_simulator.rs

```diff
@@ -34,41 +41,125 @@ use crate::c_oracle_header::{
-                .start()
-                .await;
-
-        PythSimulator {
-            program_id,
+        let mut bpf_data = read_file(find_file("pyth_oracle.so").unwrap_or_else(|| {
+            panic!("Unable to locate {}", "pyth_oracle.so");
+        }));
+
+
+        let mut program_test = ProgramTest::default();
+        let program_key = Pubkey::new_unique();
+        let programdata_key = Pubkey::new_unique();
+
+        let upgrade_authority_keypair = Keypair::new();
+
+        let program_deserialized = UpgradeableLoaderState::Program {
+            programdata_address: programdata_key,
+        };
+        let programdata_deserialized = UpgradeableLoaderState::ProgramData {
+            slot:                      1,
+            upgrade_authority_address: Some(upgrade_authority_keypair.pubkey()),
+        };
+
+        // Program contains a pointer to progradata
+        let program_vec = bincode::serialize(&program_deserialized).unwrap();
+        // Programdata contains a header and the binary of the program
+        let mut programdata_vec = bincode::serialize(&programdata_deserialized).unwrap();
+        programdata_vec.append(&mut bpf_data);
+
+        let program_account = Account {
+            lamports:   Rent::default().minimum_balance(program_vec.len()),
+            data:       program_vec,
+            owner:      bpf_loader_upgradeable::ID,
+            executable: true,
+            rent_epoch: Epoch::default(),
+        };
+        let programdata_account = Account {
+            lamports:   Rent::default().minimum_balance(programdata_vec.len()),
+            data:       programdata_vec,
+            owner:      bpf_loader_upgradeable::ID,
+            executable: false,
+            rent_epoch: Epoch::default(),
+        };
+
+        // Add to both accounts to program test, now the program is deploy as upgradable
+        program_test.add_account(program_key, program_account);
+        program_test.add_account(programdata_key, programdata_account);
+
+
+        // Start validator
+        let (mut banks_client, payer, mut recent_blockhash) = program_test.start().await;
+
+        // Transfer money to upgrade_authority so it can call the instructions
+        let instruction = system_instruction::transfer(
+            &payer.pubkey(),
+            &upgrade_authority_keypair.pubkey(),
+            1000 * LAMPORTS_PER_SOL,
+        );
+
+        let mut transaction = Transaction::new_with_payer(&[instruction], Some(&payer.pubkey()));
+
+        let blockhash = banks_client
+            .get_new_latest_blockhash(&recent_blockhash)
+            .await
+            .unwrap();
+
+        recent_blockhash = blockhash;
+
+        transaction.partial_sign(&[&payer], recent_blockhash);
+
+        banks_client.process_transaction(transaction).await.unwrap();
+
+        let genesis_keypair = Keypair::from_bytes(&payer.to_bytes()).unwrap(); // Hack to duplicate a keypair
```

Comment:
No very happy about this. I think Keypair doesn't have a Copy trait because solana doesn't want people to copy keypairs.